Environment_Engine/Compute/OpeningsFromGlazingRatio.cs

```diff
@@ -0,0 +1,98 @@
+﻿using BH.Engine.Geometry;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using BH.oM.Environment.Configuration;
+using BH.oM.Environment.Elements;
+using BH.oM.Geometry;
+using ICSharpCode.Decompiler.IL.Patterns;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+
+namespace BH.Engine.Environment
+{
+    public static partial class Compute
+    {
+        [Description("Creates openings for input Panels at the locations provided.\n The total area of the openings is equal to the total area of the external panels, multiplied by the glazing ratio.")]
+        [Input("panelsAsSpaces", "Panels as spaces - A collection of Environment Panels which will be used to identify the host panel for the opening from the provided location point.")]
+        [Input("glazingLocations", "The point in 3D space corresponding to the desired locations of the openings.")]
+        [Input("glazingRatio", "Ratio of total external panel surface area and glazed area. Ratio as decimal {0-1}.")]
+        [Input("panelsToIgnore", "Optional input for selecting a collection of Environment Panels to be ignored in the calculation")]
+        [Input("sillHeight", "Optional input for defining the distance between the base of the panel and the bottom of the opening - default: 0.5m.")]
+        [Input("openingHeight", "Optional input for defining the height of opening - default: 1.2m.")]
+        [Input("openingType", "Optional input for defining the opening type of the output Openings, can be either 'Glazing' or 'Door' - default: Glazing.")]
+        [Output("openings", "Returns the openings.")]
+        public static List<Opening> OpeningsFromGlazingRatio(
+            List<List<Panel>> panelsAsSpaces,
+            List<Point> glazingLocations,
+            double glazingRatio,
+            List<Panel> panelsToIgnore = null,
+            double sillHeight = 0.5,
+            double openingHeight = 1.2,
+            OpeningType openingType = OpeningType.Glazing
+            ) 
+        {
+            switch (openingType)
+            {
+                case OpeningType.Undefined:
+                case OpeningType.Door:
+                case OpeningType.Glazing:
+                    break;
+                default:
+                    BH.Engine.Base.Compute.RecordError($"Only openings of type 'Glazing' and 'Door' are supported.");
+                    return new List<Opening>();
+            }
+
+            List<Panel> wallPanels = panelsAsSpaces.SelectMany(x => x).Distinct().ToList();
```

Comment:
If you ultimately want a flat list of unique panels, why not ask the user to provide that rather than providing the panels as spaces? We could have another method which takes in the panels as spaces which then flattens the list for the user and feeds to this method if we really want.